### PR TITLE
Rebuild pinned environments due to `broken` pandas 0.25.0

### PR DESCRIPTION
**Pull request by leehart**
Ref: #59 

~- [x] Reinstate environment.yml~ [red herring]
- [x] Wipe old pinned environment files (linux was breaking)
- [x] Re-add pinned environment files from Travis logs
- [x] Confirm CI passes with the new pinned files.

### CODE REVIEW

**Comment by leehart** [comment]
@hardingnj Ready to review again. I realised I could use this PR to also fix/update the pinned files, as per the read-me.

**Comment by leehart** [comment]
Closing because the AWOL `environment.yml` was a red herring, and we plan to rebuild the pinned files without the build strings (e.g. `py36hb3f55d8_0`) while we investigate why `pandas==0.25.0=py36hb3f55d8_0` became unavailable.

**Comment by leehart** [comment]
Re-opening because the pinned files should still be good to go.
- [ ] update the README wrt environment.yml and requirements-*.txt files

**Comment by leehart** [comment]
@hardingnj @alimanfoo ready for review again (stale review auto dismissed)

Judging from the pandas-feedstock comments, there seems a chance that pandas `0.25.0` might be restored from `broken` to `main` within days, if that's even possible, and this PR will then become unnecessary (except for the README update), and downstream churn avoided.

**Comment by leehart** [comment]
Possibly just my machine, but when using this environment, a notebook that previously ran OK now reports `ModuleNotFoundError: No module named 'yaml'`

**Comment by alimanfoo** [comment]
> Judging from the pandas-feedstock comments, there seems a chance that pandas `0.25.0` might be restored from `broken` to `main` within days, if that's even possible, and this PR will then become unnecessary (except for the README update), and downstream churn avoided.

Given the number of bugs fixed in pandas 0.25.1, it would probably be good to upgrade anyway.

**Comment by alimanfoo** [comment]
> Possibly just my machine, but when using this environment, a notebook that previously ran OK now reports `ModuleNotFoundError: No module named 'yaml'`

Can you say which command generates this error? Can you provide a minimal reproducible example?

File: environment-pinned-linux.yml [hunk 1]
```diff
@@ -17,24 +17,24 @@ dependencies:
   - async_generator=1.10=py_0
   - attrs=19.1.0=py_0
   - backcall=0.1.0=py_0
-  - bcftools=1.9=ha228f0b_4
+  - bcftools=1.8=h4da6232_3
```
Comment (alimanfoo) [inline environment-pinned-linux.yml]: Hm, bcftools has somehow got downgraded in this update. I'm not too concerned about that as we rarely use bcftools, but it is unexpected.

File: environment-pinned-linux.yml [hunk 2]
```diff
@@ -69,51 +69,51 @@ dependencies:
...
+  - harfbuzz=2.4.0=h9f30f68_3
   - hdf4=4.2.13=h9a582f1_1002
-  - hdf5=1.10.4=nompi_h3c11f04_1106
+  - hdf5=1.10.5=nompi_h3c11f04_1103
   - heapdict=1.0.0=py36_1000
   - hmmlearn=0.2.1=py36h3010b51_1000
-  - holoviews=1.12.4=py_0
-  - htslib=1.9=ha228f0b_7
+  - holoviews=1.12.5=py_0
+  - htslib=1.7=0
```
Comment (alimanfoo) [inline environment-pinned-linux.yml]: Another downgrade here, possibly related to bcftools.

File: environment-pinned-linux.yml [hunk 3]
```diff
@@ -233,21 +235,21 @@ dependencies:
...
   - pyopenssl=19.0.0=py36_0
   - pyparsing=2.4.2=py_0
-  - pyproj=1.9.6=py36h516909a_1002
+  - pyproj=2.3.0=py36h2fd02e8_0
   - pyqt=5.9.2=py36hcca6a23_2
   - pyrsistent=0.15.4=py36h516909a_0
-  - pysam=0.15.3=py36h5ad169c_0
-  - pysamstats=1.1.2=py36h84994c4_3
+  - pysam=0.14.1=py36hae42fb6_1
+  - pysamstats=1.0.1=py36ha92aebf_2
```
Comment (alimanfoo) [inline environment-pinned-linux.yml]: More downgrades. This is not good, we want latest pysamstats.

File: environment-pinned-linux.yml [hunk 4]
```diff
@@ -256,24 +258,24 @@ dependencies:
...
   - rsa=3.4.2=py_1
-  - ruamel=1.0=py36_0
-  - ruamel.yaml=0.16.0=py36h516909a_0
-  - s3fs=0.3.1=py_0
+  - ruamel.yaml=0.16.5=py36h516909a_1
+  - ruamel.yaml.clib=0.1.2=py36h516909a_0
+  - s3fs=0.3.3=py_0
   - s3transfer=0.2.1=py36_0
-  - samtools=1.9=h8571acd_11
+  - samtools=1.7=1
```
Comment (alimanfoo) [inline environment-pinned-linux.yml]: Another downgrade.

**Comment by leehart** [comment]
Trying this again:
- [ ] Wipe old pinned environment files
- [ ] Re-add pinned environment files from Travis logs
- [ ] Confirm CI passes with the new pinned files.

**Comment by alimanfoo** [comment]
> There are several downgrades here, all come from bioconda channel, all maybe related to htslib somehow (I think all have a dependency on htslib). I think this is a problem, we do care about the version of pysamstats.

I think the problem is pysamstats, for some reason I cannot get the latest version (1.1.2) to install now, just doing:

```
conda create --yes -v $SOLVER_OPTS $CHANNEL_OPTS --name $CONDANAME pysamstats==1.1.2
```

...generates a conflict, where the environment variables are set as in install-conda.sh. I can't make any sense of the conflict error, but suspect it's something to do with numpy:

```
Package zlib conflicts for:
pysamstats==1.1.2 -> zlib[version='>=1.2.11,<1.3.0a0']
Package pysam conflicts for:
pysamstats==1.1.2 -> pysam[version='>=0.15.2,<0.16.0a0|>=0.15|>=0.15.2,<0.15.3.0a0|>=0.15.2,<1.0a0']
Package python conflicts for:
pysamstats==1.1.2 -> python[version='>=2.7,<2.8.0a0|>=3.5,<3.6.0a0|>=3.6,<3.7.0a0|>=3.7,<3.8.0a0']
Package numpy conflicts for:
pysamstats==1.1.2 -> numpy
Package libgcc-ng conflicts for:
pysamstats==1.1.2 -> libgcc-ng[version='>=4.9|>=7.3.0']
Package pytables conflicts for:
pysamstats==1.1.2 -> pytables
Note that strict channel priority may have removed packages required for satisfiability.
```

Maybe a new build of pysamstats is required on bioconda, although that's just a guess.

**Comment by alimanfoo** [comment]
xref https://github.com/bioconda/bioconda-recipes/pull/17159

**Comment by leehart** [comment]
It looks like Travis produced the same pinned versions the second time around too, fwiw.

Btw, I can no longer reproduce the `No module named 'yaml'` error that I was getting, so I guess my local environment wasn't set up properly after all, which means my workaround is now working.

**Comment by leehart** [comment]
I've been trying to verify that pandas `0.25.0` has been restored, which I believe it has, and that the `install-conda.sh` script runs successfully with the current master `enviroment-pinned-linux.yml` file, which I believe it does...
But I keep running into `conda not installed`, whenever I try to run the `env.sh` script to activate the environment, because `source activate $CONDANAME` fails, because (according to `conda info --envs`) there is no environment by that name (e.g. `vector-ops-c668z73`).
I've tried both with and without the MALARIAGEN_BINDER_HOME approach. Tried removing `binder/deps` and rerunning. Tried removing the pinned file and rerunning.
I'm expecting to see the CONDANAME turn up in `conda info --envs`, as previous ones did, but it appears to be called `base` instead.
In fact, I can cd to my binder submodule and `source activate base`, which I think might be a workaround.
I should probably raise a separate issue, but thought I should mention it here.

**Comment by leehart** [comment]
To follow on from the above, if I manually run `conda env create --name vector-ops-c668a73 --file binder/environment-pinned-linux.yml`, for example:
```
Collecting package metadata (repodata.json): done
Solving environment: \ 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
failed                                                                                                                                                            

UnsatisfiableError: The following specifications were found to be incompatible with each other:



Package scipy conflicts for:
seaborn==0.9.0=py_1 -> scipy[version='>=0.15.2']
[...]
Package pyasn1-modules conflicts for:
google-auth==1.6.3=py_0 -> pyasn1-modules[version='>=0.2.1']
Note that strict channel priority may have removed packages required for satisfiability.
```

...which I think leads me to the same place @alimanfoo got to, wrt `pysamstats`, etc.
So, the reason I see `conda not installed` is because `source activate $CONDANAME` fails, which is because the `conda env create [...]` line in `install-conda.sh` failed, which is because of package conflicts.

**Comment by leehart** [comment]
I wonder if packages from one channel have dependencies from another channel and so `strict channel priority` prevents packages from resolving dependencies between channels? Trying to decypher `Note that strict channel priority may have removed packages required for satisfiability.
`

**Comment by leehart** [comment]
^ but didn't this set up used to work?

**Comment by alimanfoo** [comment]
> ^ but didn't this set up used to work?

Yes, the really frustrating thing about this is that this used to work, using exactly the same conda version (4.7.11) and exactly the same environment-pinned-linux.yml file and exactly the same commands and options. Now I am finding, as you are, that attempting to create an environment from the pinned file causes conflicts. I can only assume that something has changed in the bioconda and/or conda-forge channels, but it's really hard to debug. 

My hunch is still that the problem is with bioconda. I'll try to do some more digging.

**Comment by leehart** [comment]
I've confirmed, unfortunately, that I get the same problem (or identical) when I try to use an old version of binder (`33ba4eab4`) that was contemporary with the committed notebook that I need to work on.

**Comment by alimanfoo** [comment]
OK, we have a fix that allows installation of the current pinned environment file. 

The important step is to ensure that strict channel priority is **not** being used when creating an environment from the pinned file. Strict channel priority is not needed then anyway because the pinned file includes all packages pinned to versions and build strings, which causes them to be pulled from the correct channel without strict channel priority. For some reason, strict channel priority also causes conflicts. 

Strict channel priority may previously have been configured in your ``.condarc`` file in your home directory. Open your ``~/.condarc`` file with a text editor, and if you see a line there that says ``channel_priority: strict`` then **delete that line**.

After that, you should be able to run the ``install-conda.sh`` script locally, and it will successfully create an environment from the pinned file.

@leehart please could you try this and confirm if you are able to install.

**Comment by leehart** [comment]
I can happily confirm that I am (after removing `channel_priority: strict` from my `.condarc`) that I am able to run `install-conda.sh` successfully, and activate the environment successfully, locally. :+1: :+1: 

Oddly I'm getting the `ModuleNotFoundError: No module named 'yaml'` error again when I try to run notebooks locally under the environment, but I reckon that's a lesser issue, and I'm guessing I've forgotten to do something.

**Comment by leehart** [comment]
TIP: the `ModuleNotFoundError: No module named 'yaml'` error goes away when I make sure the correct kernel is selected in the notebook! :+1:

**Comment by leehart** [comment]
Closing because pandas `0.25.0` has since been unbroken, and we can now run `install-conda.sh` successfully.
We also plan a new way of generating the pinned files, which will affect the README, so all of the file changes here (and this original issue) are now obsolete.

**Pull request closed without merging**